Format CSV output to allow for track-level audience reporting

jkeck opened the pull request:
Closes #6 

This PR starts w/ a refactor to make the `RWJReporter` class open to changing the CSV output, then swaps out the CSV Printer dependency in a separate commit.

jkeck committed:
Abstract the settings and CSV printing responsibilities into their own classes.

jkeck committed:
Remove RWJReporter.print_reports_for_dates as it was only used in tests.

jkeck committed:
Inject the printer dependency to RwjReporter.

jkeck committed:
Add a CSV printer class to merge show counts with track data.

jkeck committed:
Swap the default CSV printer dependency to RWJ Reporter.

ndushay commented on lib/rwj_reporter.rb:
```diff
@@ -1,10 +1,8 @@
-    rr = RwjReporter.new(start_date_str, end_date_str, log_file_dir)
-    rr.print_reports_for_dates(output_dir)
-  end
-
```
I thought this premature optimization might be useful, like if they forgot to do a month or needed 6 months or something.  But I'm not emotionally attached to it.

ndushay commented:
will data/track-template.csv   need updates over time?   Could there be a comment at the top explaining what it is?

ndushay commented:
spec/lib/track_meged_...   has a typo in the filename

ndushay commented on spec/lib/rwj_reporter_spec.rb:
```diff
@@ -102,19 +84,16 @@
-      it 'each channel file is sorted by show number' do
-        expect(channel1_data_array).to eq channel1_data_array.sort
-        expect(channel2_data_array).to eq channel2_data_array.sort
-      end
```
If we are still having channel files, should we still be ensuring they are sorted by show number?  Not clear to me why this test went away.

ndushay commented on spec/lib/rwj_reporter_spec.rb:
```diff
@@ -55,28 +50,15 @@
-        rr.print_reports_for_dates
-        expect(channel1_output_array).to eq channel1_output_array.sort
-        expect(channel2_output_array).to eq channel2_output_array.sort
-      end
```
If we are still having channel files, should we still be ensuring they are sorted by show number? Not clear to me why this test went away.

ndushay commented:
This seems ok to me.  I can't quite wrap my head around all the OO changes -- I've been too deep in other code, perhaps.   It's not clear to me all the refactoring is truly necessary, but it's good OO practice and good refactoring practice and it certainly isn't harming anything.

Does the README need to be updated?  Will the tracklist need updating? Will the user need to do anything differently?  Is the output different than stated?  Etc.

It would be great to have a "before" and "after" view of the output -- I am still not clear on exactly what was needed and exactly what is newly produced by these changes.   Maybe a little snippet of output expected in the class comments for the different printer implementations?

jkeck commented:
@ndushay I believe I have addressed your comments with changes.

FWIW, I'll answer some of them here for posterity.
1. It has been explained that the master spreadsheet does not change.
2. Since we use the master spreadsheet as our template, we can expect that is sorted appropriately (now reflected in a class comment on the CSVPrinter class).
3. The new CSV output has been documented as a class comment on the CSVPrinter class.

## Example Output

### Old

This simply gave us the show number as the first column in the CSV, and the count as the second number.

```
6,1
7,2
8,3
```

### New

This output gives the show number as the first column, the track number as the 2nd column, and the count as the 3rd number (note that the count is intentionally repeated for each track number as we do not have track-level data)

```
6,1,0
6,2,0
6,3,0
7,1,2
7,2,2
7,3,2
7,4,2
8,1,3
8,2,3
8,3,3
```

ndushay commented:
I feel better now.  👍  to merge.

jmartin-sul commented:
i did not have a chance to thoroughly read through the whole thing, and did not have a chance to figure out whether i share one of @ndushay's concerns ("If we are still having channel files, should we still be ensuring they are sorted by show number? Not clear to me why this test went away.").

still, i did look through it as individual commits, and each of those seems reasonable and well tested to me.  so i am 👍, but hesitant to merge myself (since i can't spend more time on this before i get back wednesday).

jkeck commented:
WRT @jmartin-sul's comment about the show numbers being sorted:

The sort order is already pre-determined by the track template.   Since the track template comes from the master spreadsheet (which the resulting CSV will be copied into), we **do not** want to do any sorting of the data as it would make the resulting CSV potentially not line up with the master spreadsheet from which the data originated.
